src/approaches/grammar_search_invention_approach.py

```diff
@@ -0,0 +1,198 @@
+    with the candidates proposed from a grammar.
+    """
+    def __init__(self, simulator: Callable[[State, Action], State],
+                 all_predicates: Set[Predicate],
+                 initial_options: Set[ParameterizedOption],
+                 types: Set[Type],
+                 action_space: Box,
+                 train_tasks: List[Task]) -> None:
+        super().__init__(simulator, all_predicates, initial_options,
+                         types, action_space, train_tasks)
+        self._learned_predicates: Set[Predicate] = set()
+        self._num_inventions = 0
+
+    def _get_current_predicates(self) -> Set[Predicate]:
+        return self._initial_predicates | self._learned_predicates
+
+    def learn_from_offline_dataset(self, dataset: Dataset) -> None:
+        # Generate a candidate set of predicates.
+        grammar = _create_grammar(CFG.grammar_search_grammar_name, self._types,
+                                  dataset)
+        candidates = grammar.generate(max_num=CFG.grammar_search_max_predicates)
+        # Apply the candidate predicates to the data.
+        atom_dataset = utils.create_ground_atom_dataset(dataset,
+            candidates | self._initial_predicates)
+        # Select a subset of the candidates to keep.
+        self._learned_predicates = self._select_predicates_to_keep(candidates,
+            atom_dataset)
+        # Finally, learn NSRTs via superclass, using all the kept predicates.
+        self._learn_nsrts(dataset)
+
+    def _select_predicates_to_keep(self, candidates: Set[Predicate],
+                                   atom_dataset: List[GroundAtomTrajectory]
+                                   ) -> Set[Predicate]:
+        # Perform a greedy search over predicate sets.
+        # Successively consider smaller predicate sets.
+        def _get_successors(s: FrozenSet[Predicate]
+                ) -> Iterator[Tuple[None, FrozenSet[Predicate], float]]:
+            for predicate in sorted(s):  # sorting for determinism
+                # Actions not needed. Frozensets for hashing.
+                yield (None, frozenset(s - {predicate}), 1.)
+
+        # The heuristic is where the action happens...
+        def _heuristic(s: FrozenSet[Predicate]) -> float:
+            # Relearn operators with the current predicates.
+            kept_preds = s | self._initial_predicates
+            pruned_atom_data = utils.prune_ground_atom_dataset(atom_dataset,
+                                                               kept_preds)
+            segments = [seg for traj in pruned_atom_data
+                        for seg in segment_trajectory(traj)]
+            strips_ops, _ = learn_strips_operators(segments)
+            # Score based on how well the operators fit the data.
+            num_true_positives, num_false_positives = \
+                _count_positives_for_ops(strips_ops, pruned_atom_data)
+            # Also add a size penalty.
```

Comment:
Should the size of the predicate set matter too? Or maybe actually, the generate() function should return pairs of <predicate, cost> and we look at the total cost of a predicate set? This would make sense if you imagine some program search algorithm implementing generate()